File: 287-find-the-duplicate-number/find-the-duplicate-number.java
class Solution {
    // the one already not in correct position and if there exists one element in correct position we can just return it;
    public int findDuplicate(int[] nums) {
        int i = 0;
        while(i<nums.length)
        {
            if(nums[i]!=i){
            int pos = nums[i];
            if(nums[i]!=nums[pos])
            {
                swap(nums,i,pos);
            }
            else{
                return nums[i];
            }
          }
          else i++;
        }
        for(i=0;i<nums.length;i++)
        {
            if(nums[i]!=1) return nums[i];
        }
        return -1;
    }
    public void swap(int[] n,int i,int j)
    {
        int k = n[i];
        n[i]=n[j];
        n[j]=k;
        return;
    }
}

File: 287-find-the-duplicate-number/find-the-duplicate-number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int i =0;
        while(i<nums.size())
        {
            if(nums[i]!=i)
            {
                int pos = nums[i];
                if(nums[pos]!=nums[i])
                {
                    swap(nums,i,pos);
                }
                else return nums[i];
                // since already exists in correct position then the number is duplicate
            }
            else {
                i++;
            }
        }
        return -1;
    }
    void swap(vector<int>& nums, int i, int j)
    {
        int k = nums[i];
        nums[i]=nums[j];
        nums[j]=k;
        return;
    }
};